src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs

```diff
@@ -467,5 +468,68 @@ private void CheckFieldsPropertiesMethods(Type type, string[] expectedFields, st
         }
 
         // TODO: add test guarding against (de)serializing static classes.
+
+        [Fact]
+        public void TestMultipleDefinitions()
+        {
+            // Adding a dependency to an assembly that has internal definitions of public types
+            // should not result in a collision and break generation.
+            // This verifies the usage of GetBestTypeByMetadataName() instead of GetTypeByMetadataName().
+            var referencedSource = @"
+                namespace System.Text.Json.Serialization
+                {
+                    internal class JsonSerializerContext { }
+                    internal class JsonSerializableAttribute { }
+                    internal class JsonSourceGenerationOptionsAttribute { }
+                }";
+
+            // Compile the referenced assembly first.
+            Compilation referencedCompilation = CompilationHelper.CreateCompilation(referencedSource);
+
+            // Obtain the image of the referenced assembly.
+            byte[] referencedImage;
+            using (MemoryStream ms = new MemoryStream())
+            {
+                var emitResult = referencedCompilation.Emit(ms);
+                if (!emitResult.Success)
+                {
+                    throw new InvalidOperationException();
+                }
+                referencedImage = ms.ToArray();
```

Comment:
Very minor nit: this could have been a helper method.

Reply:
Thanks; left as-is. Note there are other tests that have the same pattern.